core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala

```diff
@@ -358,6 +376,119 @@ private[mesos] trait MesosSchedulerUtils extends Logging {
   }
 
   /**
+   * Checks executor ports if they are within some range of the offered list of ports ranges,
+   *
+   * @param conf the Spark Config
+   * @param ports the list of ports to check
+   * @return true if ports are within range false otherwise
+   */
+  protected def checkPorts(conf: SparkConf, ports: List[(Long, Long)]): Boolean = {
+
+    def checkIfInRange(port: Long, ps: List[(Long, Long)]): Boolean = {
+      ps.exists(r => r._1 <= port & r._2 >= port)
+    }
+
+    val portsToCheck = nonZeroPortValuesFromConfig(conf)
+    val withinRange = portsToCheck.forall(p => checkIfInRange(p, ports))
+    // make sure we have enough ports to allocate per offer
+    ports.map(r => r._2 - r._1 + 1).sum >= portsToCheck.size && withinRange
+  }
+
+  /**
+   * Partitions port resources.
+   *
+   * @param requestedPorts non-zero ports to assign
+   * @param offeredResources the resources offered
+   * @return resources left, port resources to be used.
+   */
+  def partitionPortResources(requestedPorts: List[Long], offeredResources: List[Resource])
+    : (List[Resource], List[Resource]) = {
+    if (requestedPorts.isEmpty) {
+     return (offeredResources, List[Resource]())
+    }
+    // partition port offers
+    val (resourcesWithoutPorts, portResources) = filterPortResources(offeredResources)
+
+    val offeredPortRanges = getRangeResourceWithRoleInfo(portResources.asJava, "ports")
+
+    val portsAndRoles = requestedPorts.
+      map(x => (x, findPortAndGetAssignedRangeRole(x, offeredPortRanges)))
+
+    val assignedPortResources = createResourcesFromPorts(portsAndRoles)
+
+    // ignore non-assigned port resources, they will be declined implicitly by mesos
+    // no need for splitting port resources.
+    (resourcesWithoutPorts, assignedPortResources)
+  }
+
+  val managedPortNames = List("spark.executor.port", "spark.blockManager.port")
+
+  /**
+   * The values of the non-zero ports to be used by the executor process.
+   * @param conf the spark config to use
+   * @return the ono-zero values of the ports
+   */
+  def nonZeroPortValuesFromConfig(conf: SparkConf): List[Long] = {
+    managedPortNames.map(conf.getLong(_, 0)).filter( _ != 0)
+  }
+
+  /** Creates a mesos resource for a specific port number. */
+  private def createResourcesFromPorts(nonZero: List[(Long, String)]) : List[Resource] = {
+    nonZero.flatMap{port => createMesosPortResource(List((port._1, port._1)), Some(port._2))}
+  }
+
+  /**
+   * A resource can have multiple values in the offer since it can either be from
+   * a specific role or wildcard.
+   * Extract role info and port range for every port resource in the offer.
+   */
+  private def getRangeResourceWithRoleInfo(resources: JList[Resource], name: String)
```

Comment:
Given that we're only using this returned data structure once now, I don't think it's worth the added indirection.  Can you remove this, and just pass the raw resources around?